src/Proto3/Suite/DotProto/Generate/Common.hs

```diff
@@ -0,0 +1,358 @@
+  case (prev, cur) of
+    (Just '_', Just x)
+      | isAlpha x        -> pure (toUpper x)
+    (Just '_', Nothing)  -> pure '_'
+    (Just '_', Just '_') -> pure '_'
+    (_, Just '_')        -> empty
+    (_, Just x)          -> pure x
+    (_, _)               -> empty
+
+typeLikeName :: MonadError CompileError m => String -> m String
+typeLikeName ident@(c:cs)
+  | isUpper c = pure (camelCased ident)
+  | isLower c = pure (camelCased (toUpper c : cs))
+  | '_'  == c = pure (camelCased ('X':ident))
+typeLikeName ident = invalidTypeNameError ident
+
+fieldLikeName :: String -> String
+fieldLikeName ident@(c:_)
+  | isUpper c = let (prefix, suffix) = span isUpper ident
+                in map toLower prefix ++ suffix
+fieldLikeName ident = ident
+
+prefixedEnumFieldName :: String -> String -> String
+prefixedEnumFieldName enumName fieldName = enumName <> fieldName
+
+prefixedConName :: MonadError CompileError m => String -> String -> m String
+prefixedConName msgName conName = (msgName ++) <$> typeLikeName conName
+
+-- TODO: This should be ~:: MessageName -> FieldName -> ...; same elsewhere, the
+-- String types are a bit of a hassle.
+prefixedFieldName :: MonadError CompileError m => String -> String -> m String
+prefixedFieldName msgName fieldName = (fieldLikeName msgName ++) <$> typeLikeName fieldName
+
+dpIdentUnqualName :: MonadError CompileError m => DotProtoIdentifier -> m String
+dpIdentUnqualName (Single name)       = pure name
+dpIdentUnqualName (Dots (Path names)) = pure (last names)
+dpIdentUnqualName (Qualified _ next)  = dpIdentUnqualName next
+dpIdentUnqualName Anonymous           = internalError "dpIdentUnqualName: Anonymous"
+
+dpIdentQualName :: MonadError CompileError m => DotProtoIdentifier -> m String
+dpIdentQualName (Single name)       = pure name
+dpIdentQualName (Dots (Path names)) = pure (intercalate "." names)
+dpIdentQualName (Qualified _ _)     = internalError "dpIdentQualName: Qualified"
+dpIdentQualName Anonymous           = internalError "dpIdentQualName: Anonymous"
+
+-- | Given a 'DotProtoIdentifier' for the parent type and the unqualified name
+-- of this type, generate the corresponding Haskell name
+nestedTypeName :: MonadError CompileError m => DotProtoIdentifier -> String -> m String
+nestedTypeName Anonymous             nm = typeLikeName nm
+nestedTypeName (Single parent)       nm = intercalate "_" <$> traverse typeLikeName [parent, nm]
+nestedTypeName (Dots (Path parents)) nm = intercalate "_" . (<> [nm]) <$> traverse typeLikeName parents
+nestedTypeName (Qualified {})        _  = internalError "nestedTypeName: Qualified"
+
+qualifiedMessageName :: MonadError CompileError m => DotProtoIdentifier -> DotProtoIdentifier -> m String
+qualifiedMessageName parentIdent msgIdent = nestedTypeName parentIdent =<< dpIdentUnqualName msgIdent
+
+--------------------------------------------------------------------------------
+--
+-- ** Codegen bookkeeping helpers
+--
+
+-- | Bookeeping for qualified fields
```

Comment:
I think a lot of this can be removed, it's basically the same as `AST.DotProtoField`